File: backend/src/services/audit.service.ts
import type { Prisma } from '@prisma/client'
import { prisma } from '../config/database'
import { logger } from '../utils/logger'

export type AuditAction =
  | 'LOGIN'
  | 'LOGOUT'
  | 'REGISTER'
  | 'CREATE_COMPLAINT'
  | 'UPDATE_COMPLAINT'
  | 'ASSIGN_COMPLAINT'
  | 'CHANGE_STATUS'
  | 'CHANGE_PRIORITY'
  | 'RESOLVE_COMPLAINT'
  | 'REOPEN_COMPLAINT'
  | 'ESCALATE_COMPLAINT'
  | 'CREATE_CATEGORY'
  | 'UPDATE_CATEGORY'
  | 'DELETE_CATEGORY'
  | 'CREATE_DEPARTMENT'
  | 'UPDATE_DEPARTMENT'
  | 'DELETE_DEPARTMENT'
  | 'UPDATE_SLA'
  | 'DELETE_COMPLAINT'
  | 'CREATE_USER'
  | 'UPDATE_USER'
  | 'DELETE_USER'

interface AuditEntry {
  userId?: string | null
  action: AuditAction
  entityType: string
  entityId?: string | null
  oldValue?: unknown
  newValue?: unknown
  ipAddress?: string | null
  userAgent?: string | null
}

/**
 * Writes an audit log row. Failures are swallowed (and logged) so audit
 * capture can never break the primary request.
 */
export async function recordAudit(entry: AuditEntry): Promise<void> {
  try {
    const data: Prisma.AuditLogUncheckedCreateInput = {
      userId: entry.userId ?? null,
      action: entry.action,
      entityType: entry.entityType,
      entityId: entry.entityId ?? null,
      oldValue: entry.oldValue !== undefined ? (entry.oldValue as Prisma.InputJsonValue) : undefined,
      newValue: entry.newValue !== undefined ? (entry.newValue as Prisma.InputJsonValue) : undefined,
      ipAddress: entry.ipAddress ?? null,
      userAgent: entry.userAgent ?? null,
    }
    await prisma.auditLog.create({ data })
  } catch (err) {
    logger.warn({
      message: 'Failed to record audit log entry',
      action: entry.action,
      entityType: entry.entityType,
      error: err,
    })
  }
}

File: backend/app/services/audit_service.py
from typing import Optional, Any
from sqlalchemy.orm import Session
from app.models.audit import AuditLog

def record_audit(
    db: Session,
    action: str,
    entity_type: str,
    entity_id: Optional[str] = None,
    user_id: Optional[str] = None,
    old_value: Optional[Any] = None,
    new_value: Optional[Any] = None,
    ip_address: Optional[str] = None,
    user_agent: Optional[str] = None,
) -> AuditLog:
    log = AuditLog(
        userId=user_id,
        action=action,
        entityType=entity_type,
        entityId=entity_id,
        oldValue=old_value,
        newValue=new_value,
        ipAddress=ip_address,
        userAgent=user_agent,
    )
    db.add(log)
    db.commit()
    db.refresh(log)
    return log
